msal/application.py

```diff
@@ -1376,11 +1379,11 @@ def _acquire_token_silent_from_cache_and_possibly_refresh_it(
                 query["key_id"] = key_id
             now = time.time()
             refresh_reason = msal.telemetry.AT_ABSENT
-            for entry in self.token_cache._find(  # It returns a generator
+            for entry in self.token_cache.search(  # We want a generator here for performance, so that we can break early in cache-hit
                 self.token_cache.CredentialType.ACCESS_TOKEN,
                 target=scopes,
                 query=query,
-            ):  # Note that _find() holds a lock during this for loop;
+            ):  # Note that search() holds a lock during this for loop;
```

Comment:
`search()` uses a `RLock` which only prevents the search logic from being executed by multiple threads:

https://github.com/AzureAD/microsoft-authentication-library-for-python/blob/c4152d2920f16c0416ebcfa1b88c34d862db56e3/msal/token_cache.py#L142-L151

However, it doesn't prevent deleting an element from the token cache:

```py
import threading

mylist = [1, 2, 3]
lock = threading.RLock()


def get_next():
    with lock:
        for item in mylist:
            yield item


for i in get_next():
    print(i)
    print(mylist)
    mylist.remove(i)
    print(mylist)
```

Output:

```
1
[1, 2, 3]
[2, 3]
3
[2, 3]
[2]
```